README.md
# MATLAB Tensor Toolbox, Version 3.0-dev

Tensors (also known as multidimensional arrays or N-way arrays) are used in a 
variety of applications ranging from chemometrics to network analysis. 
The Tensor Toolbox provides the following classes for manipulating dense, 
sparse, and structured tensors using MATLAB's object-oriented features:

* `tensor` - Dense tensors, extending MATLAB's native mutlidimensional array capabilities.
* `sptensor` - Sparse tensors, only stores the nonzeros and their indices.
* `symtensor` - Symmetric tensor, only stores the unique entries.
* `ttensor` - Tucker decomposed tensor, stored as a core and factor matrices.
* `ktensor` - Kruskal decomposed tensor, stored as weight and factor matrices.
* `symktensor` - Kruskal decomposed _symmetric_ tensor, stored as weight and factor matrix.
* `sumtensor` - Sum of different types of tensors, never formed explicitly.
* `tenmat` - Tensor as a matrix, with extra information so that it can be converted back into a tensor.
* `sptenmat` - Store an sptensor as a sparse matrix in coordinate format, with extra information so that it can be converted back into an sptensor.

The Tensor Toolbox for MATLAB is open source, but we ask that you please cite 
the appropriate references ([listed below](#how-to-cite)) so that we can continue to show the 
relevance of this work. Your contributions are warmly welcomed as well; please 
see the [contribution guide](CONTRIBUTION_GUIDE.md). 
Previous contributors are listed in [contributors](CONTRIBUTORS.md). 
Full details of the license can be found in [license](LICENSE.txt).

### What's new in Version 3.0?

Version 3.0 adds

* New classes and functions for symmetric tensors: `symtensor`, `symktensor`, `cp_sym`
* New class for sums of different tensor types: `sumtensor`
* Function to compute HOSVD and ST-HOSVD: `hosvd`

We have also fixed many bugs. View the [RELEASE_NOTES.txt](RELEASE_NOTES.txt) file for details.

## How to cite

If you have used the Tensor Toolbox in your work in any way, 
please cite the software itself along with at least one publication or preprint. 
Thanks very much for your support.

__General software reference, should always be cited:__
Brett W. Bader, Tamara G. Kolda and others. 
MATLAB Tensor Toolbox Version 3.0-dev, 
Available online, August 2017. 
URL: https://gitlab.com/tensors/tensor_toolbox. 
_Consider adding the short hash for the exact version that was used. 
If you clone the repository, use the command 
`git log --pretty=format:'%h' -n 1`. 
If you download, the long hash is baked into the filename, but you need only use 
the first 8 characters._ 

``` bibtex
@misc{TTB_Software,
  author = {Brett W. Bader and Tamara G. Kolda and others},
  title = {MATLAB Tensor Toolbox Version 3.0-dev},
  howpublished = {Available online},
  month = aug,
  year = {2017},
  url = {https://gitlab.com/tensors/tensor_toolbox}
}
```

__Default citation for dense computations:__
B. W. Bader and T. G. Kolda.
Algorithm 862: MATLAB tensor classes for fast algorithm prototyping,
ACM Transactions on Mathematical Software 32(4):635-653, December 2006.
DOI: 10.1145/1186785.1186794. 

``` bibtex
@article{TTB_Dense,
  author = {Brett W. Bader and Tamara G. Kolda},
  title = {Algorithm 862: {MATLAB} tensor classes for fast algorithm prototyping},
  journal = {ACM Transactions on Mathematical Software},
  month = dec,
  year = {2006},
  volume = {32},
  number = {4},
  pages = {635--653},
  doi = {10.1145/1186785.1186794}
}
```

__Default citation for sparse computations:__
B. W. Bader and T. G. Kolda.
Efficient MATLAB computations with sparse and factored tensors,
SIAM Journal on Scientific Computing 30(1):205-231, December 2007. DOI: 10.1137/060676489. 

``` bibtex
@article{TTB_Sparse,
  author = {Brett W. Bader and Tamara G. Kolda},
  title = {Efficient {MATLAB} computations with sparse and factored tensors},
  journal = {SIAM Journal on Scientific Computing},
  month = dec,
  year = {2007},
  volume = {30},
  number = {1},
  pages = {205--231},
  doi = {10.1137/060676489}
}
```

__Citation for all-at-once optimization for CP factorization (`cp_opt`):__
E. Acar, D. M. Dunlavy and T. G. Kolda.
A Scalable Optimization Approach for Fitting Canonical Tensor Decompositions,
Journal of Chemometrics 25(2):67-86, February 2011. DOI: 10.1002/cem.1335. 

``` bibtex
@article{TTB_CPOPT,
  author = {Evrim Acar and Daniel M. Dunlavy and Tamara G. Kolda},
  title = {A Scalable Optimization Approach for Fitting Canonical Tensor Decompositions},
  journal = {Journal of Chemometrics},
  month = feb,
  year = {2011},
  volume = {25},
  number = {2},
  pages = {67--86},
  doi = {10.1002/cem.1335}
}
```

__Citation for CP factorization with missing data (`cp_wopt`):__
E. Acar, D. M. Dunlavy, T. G. Kolda and M. Mørup.
Scalable Tensor Factorizations for Incomplete Data,
Chemometrics and Intelligent Laboratory Systems 106(1):41-56, March 2011.
DOI: 10.1016/j.chemolab.2010.08.004. 

``` bibtex
@article{TTB_CPWOPT,
  author = {Evrim Acar and Daniel M. Dunlavy and Tamara G. Kolda and Morten M{\o}rup},
  title = {Scalable Tensor Factorizations for Incomplete Data},
  journal = {Chemometrics and Intelligent Laboratory Systems},
  month = mar,
  year = {2011},
  volume = {106},
  number = {1},
  pages = {41--56},
  doi = {10.1016/j.chemolab.2010.08.004}
}
```

__Citation for Shifted Symmetric Higher-Order Power Method for Tensor Eigenvalues (`eig_sshopm`):__
T. G. Kolda and J. R. Mayo.
Shifted Power Method for Computing Tensor Eigenpairs,
SIAM Journal on Matrix Analysis and Applications 32(4):1095-1124, October 2011. DOI: 10.1137/100801482. 

``` bibtex
@article{TTB_SSHOPM,
  author = {Tamara G. Kolda and Jackson R. Mayo},
  title = {Shifted Power Method for Computing Tensor Eigenpairs},
  journal = {SIAM Journal on Matrix Analysis and Applications},
  month = oct,
  year = {2011},
  volume = {32},
  number = {4},
  pages = {1095-1124},
  doi = {10.1137/100801482}
}
```

__Citation for Method for Generalized Tensor Eigenvalues (`eig_geap`):__
T. G. Kolda and J. R. Mayo.
An Adaptive Shifted Power Method for Computing Generalized Tensor Eigenpairs,
SIAM Journal on Matrix Analysis and Applications 35(4):1563-1581, December 2014.
DOI: 10.1137/100801482. 

``` bibtex
@Article{TTB_EIGGEAP,
  title                    = {An Adaptive Shifted Power Method for Computing Generalized Tensor Eigenpairs},
  author                   = {Tamara G. Kolda and Jackson R. Mayo},
  doi                      = {10.1137/140951758},
  journal                  = {SIAM Journal on Matrix Analysis and Applications},
  number                   = {4},
  volume                   = {35},
  year                     = {2014},
  month                    = dec,
  pages                    = {1563--1581},
  url                      = {http://epubs.siam.org/toc/sjmael/35/4},
}
```

__Citations for on Poisson Tensor Factorization (`cp_apr`)__
1. E. C. Chi and T. G. Kolda.
On Tensors, Sparsity, and Nonnegative Factorizations,
SIAM Journal on Matrix Analysis and Applications 33(4):1272-1299, December 2012.
2. S. Hansen, T. Plantenga and T. G. Kolda.
Newton-Based Optimization for Kullback-Leibler Nonnegative Tensor Factorizations,
Optimization Methods and Software 30(5):1002-1029, April 2015. 

``` bibtex
@Article{TTB_CPAPR,
  title                    = {On Tensors, Sparsity, and Nonnegative Factorizations},
  author                   = {Eric C. Chi and Tamara G. Kolda},
  doi                      = {10.1137/110859063},
  journal                  = {SIAM Journal on Matrix Analysis and Applications},
  number                   = {4},
  volume                   = {33},
  year                     = {2012},
  month                    = dec,
  pages                    = {1272-1299},
}

@Article{TTB_CPAPRB,
  author = {Samantha Hansen and Todd Plantenga and Tamara G. Kolda}, 
  title = {Newton-Based Optimization for {Kullback-Leibler} Nonnegative Tensor Factorizations}, 
  journal = {Optimization Methods and Software}, 
  volume = {30}, 
  number = {5}, 
  pages = {1002-1029},
  month = {April}, 
  year = {2015},
  doi = {10.1080/10556788.2015.1009977},
} 
```
__Citation for Symmetric CP (`cp_sym`):__
T. G. Kolda, 
Numerical Optimization for Symmetric Tensor Decomposition, 
Mathematical Programming B 151(1):225-248, April 2015, doi:10.1007/s10107-015-0895-0
``` bibtex
@article{TTB_CPSYM,  
author = {Tamara G. Kolda}, 
title = {Numerical Optimization for Symmetric Tensor Decomposition}, 
journal = {Mathematical Programming B}, 
volume = {151}, 
number = {1}, 
pages = {225-248}, 
month = apr, 
year = {2015},
doi = {10.1007/s10107-015-0895-0},
}
```

__Citation for CP with randomized least squares (`cp_rals`):__
C. Battaglino, G. Ballard and T. G. Kolda, 
A Practical Randomized CP Tensor Decomposition, 
arXiv:1701.06600, January 2017

```bibtex
@misc{TTB_CPRALS,  
author = {Casey Battaglino and Grey Ballard and Tamara G. Kolda}, 
title = {A Practical Randomized {CP} Tensor Decomposition}, 
month = jan, 
year = {2017},
eprint = {1701.06600},
eprintclass = {cs.NA},
}
```


__Citation for Memory-Efficient Tucker (`tucker_me` and `ttm_me`):__
T. G. Kolda and J. Sun.
Scalable Tensor Decompositions for Multi-aspect Data Mining,
ICDM 2008: Proceedings of the 8th IEEE International Conference on Data Mining,
pp. 363-372, December 2008. DOI: 10.1109/ICDM.2008.89.
_This code is no longer included with the toolbox but can be found in 
Version 2.6._

``` bibtex
@inproceedings{TTB_MET,
  author = {Tamara G. Kolda and Jimeng Sun},
  title = {Scalable Tensor Decompositions for Multi-aspect Data Mining},
  booktitle = {ICDM 2008: Proceedings of the 8th IEEE International Conference on Data Mining},
  month = dec,
  year = {2008},
  pages = {363--372},
  doi = {10.1109/ICDM.2008.89}
}
```


## Getting started and using the software

### Download

Version 3.0-dev can be obtained here by cloning or downloading using the buttons 
above. There is not yet an official release 3.0.
Version 2.6 and earlier can be obtained 
[here](http://www.sandia.gov/~tgkolda/TensorToolbox/index-2.6.html).

### Installation
1. Unpack the files, if necessary
2. Start MATLAB
3. Within MATLAB, navigate to the `tensor_toolbox` directory and execute the following commands:
    1. `addpath(pwd)`
    2. `savepath`

### Getting help
At any time, type `help tensor_toolbox` for help on classes or functions. 
You can also find a getting started guide via MATLAB's help system. Launch help
by pressing the question mark button and look for Tensor Toolbox under supplemental
software, as highlighted in the image below.

![Navigating MATLAB Help Screen](doc/images/helpscreen.PNG "Navigating MATLAB Help Screen")

Copyright 2017, Sandia Corporation.



RELEASE_NOTES.txt
Tensor Toolbox version 3.1
by Brett W. Bader, Tamara G. Kolda, Evrim Acar, Daniel M. Dunlavy, et al.
Copyright 2019, Sandia National Laboratories.

============================================
Changes from Version 3.1 (June 4, 2019)
============================================
- Adding support for KTENSOR in EXPORT_DATA.
- Fixing bug in KTENSOR/SCORE with zero lambdas in both inputs.

============================================
Changes from Version 2.6 (February 6, 2015)
============================================
- Changed license conditions: now open source BSD license.
- New KTENSOR/VIZ function for visualizing the factors produced by the
  CP decomposition.
- Added new CP_SYM and TUCKER_SYM functions for symmetric tensor
  decompositions. Added new SYMTENSOR and SYMKTENSOR classes with
  limited functionality to support symmetric tensors.
- Added new SUMTENSOR class that works with an implicit sum of tensors
  without actually forming the result.
- Added new CP-ARLS method that does alternating *RANDOMIZED* least
  squares fitting for the CP decomposition per Battaglino et al.
- New GCP_OPT method for generalized CP.
- New CREATE_PROBLEM_BINARY method for generating problems where the
  low-rank model corresponds to the odds of a 1.
- Improve KTENSOR/FULL function.
- Added SPTENSOR/SPONES function that replaces nonzero sparse tensor
  elements with ones.
- Removed memory-efficient Tucker (met) code.
- Fixed formatting of lambda in ktensor/disp.
- Fixed type of subs in import_data for sptensor data.
- Made call to fixsigns in cp_als optional.

============================================
Changes from Version 2.5 (February 1, 2012)
============================================
Top Level
- Added new EIG_GEAP function for computing generalized tensor
  eigenpairs. Renamed SSHOPM to EIG_SSHOPM and added support for
  adaptive shift (now the default).  Renamed SSHOPMC to EIG_SSHOPMC.
- Major updates to CP_APR, including changing the default to use
  2nd-order optimization per paper of Hansen, Plantenga, & Kolda. See
  method help for more information. 
- Minor changes to CP_ALS: (1) Fixed bug in normalization step. (2)
  Updated some calculations per work of Phan Anh Huy. (3) Forced
  printing of last iteration so long as printitn > 0.
- Updated MTTKRP and KHATRIRAO, per work of Phan Anh Huy.
- Fixed bug in CREATE_PROBLEM for 'Sparse_Generation'.
- Added SPTENSOR support EXPORT_DATA and IMPORT_DATA. Added KTENSOR
  support to IMPORT_DATA.
- Updated random number generator references to the new MATLAB
  implementsion in CREATE_PROBLEM and CREATE_GUESS.
- Added instructions for adding MET to the pat in INSTALL.txt.
- Fixed function name for TT_FAC_TO_VEC per Evrim Acar bug report.
- Added new function MATRANDNORM.
- Renamed TT_CCONG to MATRANDCONG, TT_RANDORTHMATH to MATRANDORTH.
- Removed TT_ASSIGNMENT_TYPE, TT_COMBINATOR, TT_CP_W*,
  TT_CREATE_MISSING_DATA_PATTERN. 
- Modernized documentation with class support.

Class: ktensor
- Fixed SUBSREF to properly handle lists of indices.
- Only call ARRANGE from NORMALIZE if there are multiple
  components. Force ARRANGE to produce dense matrices. Fixes bug
  reported by Jason Mattax on 3/1/2012.
- Fixed comments for NORMALIZE, SUBSREF.
- Added ISSYMMETRIC and SYMMETRIZE functions.
- Constructor can now take SYMKTENSOR as an input.

Class: sptensor
- Fixed bug in SUBSASGN discovered by Sebastien Bratieres pertaining
  to empty tensors.

Class: tensor
- Fixed bug in ISSYMMETRIC with respect to groups.

Acknowledgments:
- The function @symtensor/private/multinomial.m is from Mukhtar Ullah
  and was distributed via the MATLAB file exchange.

============================================
Changes from Version 2.4 (March 22, 2010)
============================================

Top Level
- The "algorithms" directory has been eliminated. All routines are now
  at the root level, meaning that only one directory has to be added
  to the path to get all of Tensor Toolbox's standard functionality. 
- Added new CREATE_PROBLEM and CREATE_GUESS routines that can be used
  to generate test problems and initial guesses. These were first used
  at the AIM 2010 Tensor Decomposition workshop. Added TT_RANDORTHMAT,
  a helper function for creating problems.
- Added new SSHOPM and SSHOPMC code for Shifted Symmetric Higher-Order
  Power Method for computing tensor eigenpairs.
- Added new CP_APR method for Poisson Tensor Factorization via
  alternating Poisson regression, along with helper function
  tt_loglikelihood. 
- Added TENEYE to create "identity tensor".
- Helper functions for CP_OPT and CP_WOPT (like cp_fg) now have "tt_"
  prepended to their names. They are not listed in the contents files.
- Adding ability to import and export text versions of matrices and
  tensors via IMPORT_DATA/EXPORT_DATA functions.
- Making calling sequence to TENZEROS, TENRAND, and TENONES
  consistent. Now all three will take either a size array or a list,
  i.e., tenones([5 4 3]) or tenones(5,4,3) produce the same
  results. Eliminated two-argument version of tenzeros, i.e., a call
  to tenzeros(M,N) should be changed to tenzeros(N*ones(1,M)).
- Added additional comments in CP_ALS.
- Made output of CP_WOPT consistent with CP_OPT, i.e., now includes
  output of optimization method.  
- Fixed bug for empty tensor in TENONES.
- Fixed TT_IND2SUB, TT_SUB2IND to handle empty inputs.

Documentation
- Added documentation in the help browser for cp_opt, cp_wopt, cp_als,
  and sshopm.

Class: tensor
- Added SYMMETRIZE function to symmetrize a tensor and ISSYMMETRIC
  function to check if a tensor is symmetric.
- Adding new TTSV function to compute a tensor times the same vector
  in every mode. Intended for symmetric tensors and doesn't allow user
  to specify exactly which modes are skipped.
- Fixed "empty tensor" bugs in TENSOR (constructor), PERMUTE, COLLAPSE.
- Fixed "1D tensor" bug in TENMAT.
- Fixed bug with no results in FIND.
- More error checking in MTTKRP.

CLASS: sptensor

cp_opt.m
function [P, P0, output] = cp_opt(Z,R,varargin)
%CP_OPT Fits a CP model to a tensor via optimization.
%
%   K = CP_OPT(X,R) fits an R-component CANDECOMP/PARAFAC (CP) model
%   to the tensor X. The result K is a ktensor. The function being
%   optimized is F(K) = 1/2 || X - K ||^2.
%
%   K = CP_OPT(X,R,'param',value,...) specifies additional
%   parameters for the method. Specifically...
%
%   'init' - Initialization for factor matrices (default: 'randn'). This
%   can be a cell array with the initial matrices, a ktensor, or one of the
%   following strings:
%      'randn'  Randomly generated via randn function
%      'rand'   Randomly generated via rand function
%      'zeros'  All zeros
%      'nvecs'  Selected as leading left singular vectors of X(n)
%
%   'opt_options' - Optimization method options, passed as a structure.
%   Type 'help lbfgsb' to see the options. (Note that the 'opts.x0' option
%   is overwritten using the choice for 'init', above.) 
%
%   'lower'/'upper' - Lower/upper bounds, passed in as a scalar (if they
%   are all the same), vector, cell array, or ktensor (lambda values
%   ignored).  
%
%   [K, U0] = CP_OPT(...) also returns the initial guess.
%
%   [K, U0, OUT] = CP_OPT(...) also returns a structure with the
%   optimization exit flag, the final relative fit, and the full
%   output from the optimization method. The fit is defined as 
%
%      FIT = 100 * (1 - ( F(K) / F(0) )).
%
%   REFERENCE: E. Acar, D. M. Dunlavy and T. G. Kolda, A Scalable
%   Optimization Approach for Fitting Canonical Tensor Decompositions,
%   J. Chemometrics 25(2):67-86, February 2011 (doi:10.1002/cem.1335)
%
%   <a href="matlab:web(strcat('file://',...
%   fullfile(getfield(what('tensor_toolbox'),'path'),'doc','html',...
%   'cp_opt_doc.html')))">Documentation page for CP-OPT</a>
%
%   See also TENSOR, SPTENSOR, KTENSOR.
%
%MATLAB Tensor Toolbox. Copyright 2018, Sandia Corporation.


%% Error checking
% if ~isa(Z,'tensor') && ~isa(Z,'sptensor')
%     error('Z must be a tensor or a sptensor');
% end
% 
if (nargin < 2)
    error('Error: invalid input arguments');
end

%% Set parameters
params = inputParser;
params.addParameter('opt', 'lbfgsb', @(x) ismember(x,{'ncg','tn','lbfgs','lbfgsb'}));
params.addParameter('init', 'randn', @(x) (iscell(x) || isa(x, 'ktensor') || ismember(x,{'random','rand','randn','nvecs','zeros'})));
params.addParameter('lower',-Inf);
params.addParameter('upper',Inf);
params.addParameter('opt_options', '', @isstruct);
params.parse(varargin{:});

init = params.Results.init;
opt = params.Results.opt;
options = params.Results.opt_options;
lower = params.Results.lower;
upper = params.Results.upper;

use_lbfgsb = strcmp(opt,'lbfgsb');

%% Initialization
sz = size(Z);
N = length(sz);

if iscell(init)
    P0 = init;
elseif isa(init,'ktensor')
    P0 = tocell(init);
else
    P0 = cell(N,1);
    if strcmpi(init,'nvecs')
        for n=1:N
            P0{n} = nvecs(Z,n,R);
        end
    else
        for n=1:N
            P0{n} = matrandnorm(feval(init,n,R));
        end
    end
end

%% Set up lower and upper (L-BFGS-B only)

if ~use_lbfgsb && ( any(isfinite(lower)) || any(isfinite(upper)) )
    error('Cannot use lower and upper bounds without L-BFGS-B');
end

if use_lbfgsb
    lower = convert_bound(lower,sz,R);
    upper = convert_bound(upper,sz,R);
end


%% Set up optimization algorithm

if use_lbfgsb % L-BFGS-B
    if ~exist('lbfgsb','file')
        error(['CP_OPT requires L-BFGS-B function. This can be downloaded'...
            'at https://github.com/stephenbeckr/L-BFGS-B-C']);
    end
else % POBLANO
    switch (params.Results.opt)
        case 'ncg'
            fhandle = @ncg;
        case 'tn'
            fhandle = @tn;
        case 'lbfgs'
            fhandle = @lbfgs;
    end
    
    if ~exist('poblano_params','file')
        error(['CP_OPT requires Poblano Toolbox for Matlab. This can be ' ...
            'downloaded at http://software.sandia.gov/trac/poblano.']);
    end     
end

%% Set up optimization algorithm options
if isempty(options) 
    if use_lbfgsb
        options.maxIts = 10000;
        options.maxTotalIts = 50000;
        options.printEvery = 10;
    else
        options = feval(fhandle, 'defaults');
    end
end
    



%% Fit CP using CPOPT
normsqr = norm(Z)^2;
if use_lbfgsb
    opts = options;
    opts.x0 = tt_fac_to_vec(P0);    
    [xx,ff,out] = lbfgsb(@(x)tt_cp_fun(x,Z,normsqr), lower, upper, opts);
    P = ktensor(tt_cp_vec_to_fac(xx, Z));
    output.ExitMsg = out.lbfgs_message1;
    output.Fit = 100 * (1 - ff /(0.5 * normsqr));
    output.OptOut = out;
else % POBLANO
    out = feval(fhandle, @(x)tt_cp_fun(x,Z,normsqr), tt_fac_to_vec(P0), options);
    P = ktensor(tt_cp_vec_to_fac(out.X, Z));
    output.ExitFlag  = out.ExitFlag;
    output.Fit = 100 * (1 - out.F /(0.5 * normsqr));
    output.OptOut = out;
end


%% Clean up final result
% Arrange the final tensor so that the columns are normalized.
P = arrange(P);
% Fix the signs
P = fixsigns(P);


function newbound = convert_bound(bound,sz,R)

len = sum(sz)*R;

if isscalar(bound)
    newbound = bound * ones(len,1);
elseif isa(bound,'ktensor')
    newbound = tovec(bound, false);
elseif iscell(bound)
    newbound = tt_fac_to_vec(bound);
end

if ~isequal(size(newbound), [len 1])
    error('Bound is the wrong size');
end
